device/core/src/internal_client.ts

```diff
@@ -137,141 +137,162 @@ export abstract class InternalClient extends EventEmitter {
-      }
-    });
+      const retryOp = new RetryOperation(this._retryPolicy, this._maxOperationTimeout);
+      retryOp.retry((opCallback) => {
+        /*Codes_SRS_NODE_INTERNAL_CLIENT_16_021: [The ‘setTransportOptions’ method shall call the ‘setOptions’ method on the transport object.]*/
+        this._transport.setOptions(clientOptions, opCallback);
+      }, (err) => {
+        if (err) {
+          safeCallback(_callback, err);
+        } else {
+          safeCallback(_callback, null, new results.TransportConfigured());
+        }
+      });
+    }, done);
   }
 
   /**
    * Passes options to the `Client` object that can be used to configure the transport.
    * @param options   A {@link DeviceClientOptions} object.
-   * @param done      The callback to call once the options have been set.
+   * @param [done]      Optional callback to call once the options have been set.
+   * @returns {Promise<results.TransportConfigured> | void} Promise if no callback function was passed, void otherwise.
    */
-  setOptions(options: DeviceClientOptions, done?: (err?: Error, result?: results.TransportConfigured) => void): void {
-    /*Codes_SRS_NODE_INTERNAL_CLIENT_16_042: [The `setOptions` method shall throw a `ReferenceError` if the options object is falsy.]*/
-    if (!options) throw new ReferenceError('options cannot be falsy.');
-    let localOptions: DeviceClientOptions = {};
-    for (let k in options) {
-      localOptions[k] = options[k];
-    }
+  setOptions(options: DeviceClientOptions, done?: Callback<results.TransportConfigured>): Promise<results.TransportConfigured> | void {
+    return callbackToPromise((_callback) => {
+      /*Codes_SRS_NODE_INTERNAL_CLIENT_16_042: [The `setOptions` method shall throw a `ReferenceError` if the options object is falsy.]*/
+      if (!options) throw new ReferenceError('options cannot be falsy.');
+      let localOptions: DeviceClientOptions = {};
+      for (let k in options) {
+        localOptions[k] = options[k];
+      }
 
-    /*Codes_SRS_NODE_INTERNAL_CLIENT_06_001: [The `setOptions` method shall first test if the `ca` property is the name of an already existent file.  If so, it will attempt to read that file as a pem into a string value and pass the string to config object `ca` property.  Otherwise, it is assumed to be a pem string.] */
-    if (localOptions.ca) {
-      if (fs.existsSync(localOptions.ca)) {
-        localOptions.ca = fs.readFileSync(localOptions.ca, 'utf8');
+      /*Codes_SRS_NODE_INTERNAL_CLIENT_06_001: [The `setOptions` method shall first test if the `ca` property is the name of an already existent file.  If so, it will attempt to read that file as a pem into a string value and pass the string to config object `ca` property.  Otherwise, it is assumed to be a pem string.] */
+      if (localOptions.ca) {
+        if (fs.existsSync(localOptions.ca)) {
+          localOptions.ca = fs.readFileSync(localOptions.ca, 'utf8');
```

Comment:
Using readFileSync() inside of an asynchronous method is probably a bad idea. I know you didn't write this code, but the owners should probably fix this.

Reply:
Fix for this has been checked it.

---
In reply to: [219350191](https://github.com/Azure/azure-iot-sdk-node/pull/372#discussion_r219350191) [](ancestors = 219350191)

Reply:
Done.

---
In reply to: [218891214](https://github.com/Azure/azure-iot-sdk-node/pull/372#discussion_r218891214) [](ancestors = 218891214)